### DISCUSSION TITLE
Running project tests locally

### DISCUSSION THREAD

**Discussion by jrochkind** [bensheldon/good_job#631] [Q&A] [answered]
Sorry for so many issues, I've been spending a couple days investigating good_job!  I figured it would be less disruptive here than in GH Issues for this one! 

I can't successfully run tests locally. The [README just suggests](https://github.com/bensheldon/good_job#running-tests):

     bin/rspec

## DB setup

The first thing I noticed (this isn't my first Rails engine rodeo) is I actually had to first run:

    RAILS_ENV=test ./bin/rails db:create db:migrate

(Perhaps this should be added to instrutions? Or I see your CI uses `db:test:prepare`, okay better! Add to README?)

## chromedriver needed

But okay, even after doing so, I'm still getting some local errors on `main`. A bunch of failures due to:

```
            Selenium::WebDriver::Error::WebDriverError:
              Unable to find chromedriver. Please download the server from
              https://chromedriver.storage.googleapis.com/index.html and place it somewhere on your PATH.
              More info at https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver.
```

I'm not sure what to do about this. Interestingly, I do have other projects that test with selenium and chromedriver, and work, I'm not sure what the difference is.

Hm, some (not all!) of my existing local project copies include the [webdrivers](https://github.com/titusfortner/webdrivers) gem to install chromedriver -- experimentally adding that to Gemfile, that takes care of it for me! Include that gem here? Or provide alternate instructions for installing chromedriver in a way good_job test suite can find it, if you prefer?

## other failures

But then a 5 non-chromedriver-related failures seem to occur for me too when running tests locally. 

```
  1) Server modes when production async does not start GoodJob when running other commands
     Failure/Error: expect(shell.output).to include(/Current version/)

       expected ["rails aborted!\n", "ActiveRecord::NoDatabaseError: connection to server at \"::1\", port 5432 faile...not exist\n", "\n", "Tasks: TOP => db:version\n", "(See full trace by running task with --trace)\n"] to include /Current version/

  2) Server modes when development async successfully runs in the server
     Failure/Error: expect(shell.output).to include(/Listening on/)

       expected ["=> Booting Puma\n", "=> Rails 6.1.5.1 application starting in development \n", "=> Run `bin/rails s...8:in `<top (required)>'\n", "\tfrom bin/rails:4:in `require'\n", "\tfrom bin/rails:4:in `<main>'\n"] to include /Listening on/

  16) GoodJob::Lockable.advisory_lock_key when a block is passed locks that key for the bloc and then unlocks it
      Failure/Error: expect(PgLock.advisory_lock.count).to eq 1

        expected: 1
             got: 2

        (compared using ==)

  17) GoodJob::Lockable.with_advisory_lock can unlock all advisory locks on the session with `unlock_session: true`
      Failure/Error: expect(PgLock.advisory_lock.count).to eq 0

        expected: 0
             got: 1

        (compared using ==)

  18) GoodJob::Lockable.with_advisory_lock opens a block with a lock that locks and unlocks records
      Failure/Error: expect(PgLock.advisory_lock.count).to eq 0

        expected: 0
             got: 1

        (compared using ==)
```

Not sure what's up here. Am I running the tests wrong? Something wrong in my environment? (Does good_job have particular postgres version requirements I might failing? I can't find any mentioned in README).  

Just brainstorming what could possibly be making tests fail for me locally, really not sure, do you have any ideas? Are you able to run the test suite locally?

**Comment by bensheldon** [comment, marked as answer]
@jrochkind I really appreciate the issues and conversation ❤️ 

- I do TDD locally, and I intend the repo/Readme to be batteries-included for development.
- Did you run `bin/setup`. The Gem Development part of the readme was intended to be followed linearly, from development setup (running the demo/console) through testing: https://github.com/bensheldon/good_job#gem-development
- I think you're right about Chrome/Chromedriver. I myself have had trouble with webdrivers gem, but I can add a check in `bin/setup` for that dependency. Same goes for Postgres Version (though I would be surprised if you're running PG < 9.6. Compatibility documented here: https://github.com/bensheldon/good_job#compatibility

**Reply by jrochkind** [reply]
Aha, I had not noticed bin/setup, thanks!

(There is a LOT in the README, so I'm skimming for what I need, and did not think I was just yet going to need to boot the test app in development mode or use the console -- for me the FIRST thing I do is demonstrate passing test suite, and depending on the development task I might not need a console or booted dummy app.  But that's me, and and fair enough, it was there!)

## chromedriver

bin/setup DOES warn me `!!! Chrome/Chromedriver must be installed to run Rails System tests`. The thing is, I don't actually know how to do this -- I definitely have Chrome already installed! On My MacOS system will `brew install chromedriver` do it?  Unfortunately, no, when I run bin/setup again, that gets me a message "“chromedriver” can’t be opened because Apple cannot check it for malicious software. This software needs to be updated. Contact the developer for more information."

I can try reporting the problem to homebrew, if it hasn't already been reported. 

In the meantime, I'll just add `webdrivers` to my local Gemfile as an uncommitted change, to proceed. (I'm curious what problems you've had with webdrivers? If they were a long time ago, ti's possible webdrivers has improved? I have found it magically to just work on MacOS dev machines and a variety of Linux build environments, but I don't know if I've tried on Windows). 

## faraday warning?

./bin/setup ends up printing this to console a few times, I don't know if it matters?

> To use retry middleware with Faraday v2.0+, install `faraday-retry` gem

## Still test failure, alas

I'm afraid I still have test failures in dev.  And maybe that faraday warning wasn't ignorable. Can you do a `bundle update` and verify tests still pass for you locally?

```
  1) Server modes when development async successfully runs in the server
     Failure/Error: expect(shell.output).to include(/Listening on/)

       expected ["To use retry middleware with Faraday v2.0+, install `faraday-retry` gem\n", "=> Booting Puma\n", "=...8:in `<top (required)>'\n", "\tfrom bin/rails:4:in `require'\n", "\tfrom bin/rails:4:in `<main>'\n"] to include /Listening on/
       Diff:
       @@ -1,53 +1,105 @@
       -[/Listening on/]

  2) GoodJob::Lockable.with_advisory_lock opens a block with a lock that locks and unlocks records
     Failure/Error: expect(PgLock.advisory_lock.count).to eq 0

       expected: 0
            got: 1

       (compared using ==)

  3) GoodJob::Lockable.with_advisory_lock can unlock all advisory locks on the session with `unlock_session: true`
     Failure/Error: expect(PgLock.advisory_lock.count).to eq 0

       expected: 0
            got: 1

       (compared using ==)

  4) GoodJob::Lockable.advisory_lock_key when a block is passed locks that key for the bloc and then unlocks it
     Failure/Error: expect(PgLock.advisory_lock.count).to eq 1

       expected: 1
            got: 2

       (compared using ==)
```

My postgres is 14.3, so indeed that's not the issue. I wonder why I'm getting those lock errors that CI doesn't get and you aren't getting locally either. 

It does seem concerning if test suite doesn't pass for a new developer!

**Reply by bensheldon** [reply]
> Still test failure, alas

Could you check your Activity Manager / top to see if you have some zombied `ruby` processes that are still running. That there are extra Advisory Locks makes me there could be a live testing process that wasn't cleaned up from a previous run. (If that is the problem, I can write a before_suite check for that)

> chromedriver

That sounds like typical MacOS notarization problem. Here's a solution that works from the terminal, or the other answer is how to do it through System Preferences -> Security & Privacy -> General: 
https://stackoverflow.com/a/60374958

> Faraday

I'm not sure where that warning is coming from. Faraday is not an immediate dependency of GoodJob. 


> It does seem concerning if test suite doesn't pass for a new developer!

Strong agree! You're the first person to dig into this with me (which I assume is less that others haven't had problems, but that they didn't share them). Thank you 🙏🏻

**Comment by jrochkind** [comment]
> Could you check your Activity Manager / top to see if you have some zombied ruby processes that are still running.

Nothing suspicious, but, wait... this is in a postgres instance that I use for other things -- it's like the main postgres instance on my machine, which is what good job ./bin/setup created dbs in, of course.  Nothing else should be touching the `good_job_test` database, but I _do_ have other processes connected to other databases in the postgres instance, should that be a problem?

> That sounds like typical MacOS notarization problem. 

OK, thanks. It can be worked around, but... still curious why not just include webdrivers like every other project I've worked on needing chromedriver!

> I'm not sure where that warning is coming from. Faraday is not an immediate dependency of GoodJob.

Weird. Let's see if I can figure it out. That does make me wonder if somehow some other project of mine is somehow getting involved in the selenium tests too... although the faraday warning came from `./bin/setup` too... 

Faraday may not be a direct dependency, but okay, let's see why it's in my dependency tree at all...  

OK, faraday is an indirect dependency multiple times (and I should really write a tool to automate this, why hasn't anyone else, it's confusing to get it out of the Gemfile.lock)

* github_changelog_generator -> [faraday-http-cache,faraday] -> faraday
  * github_changelog_generator -> octokit -> faraday
* ~fugit -> et-orbi -> faraday~ (nope I misread)

Hmm, github_changelog_generator is a development dependency... ~I'm going to suspect `fugit` first, which I've never heard of. But why is a time parsing tool dependent on faraday! It's possible fugit/et-orbi isn't compatible with faraday 2... but why wouldn't that show up in CI, why just on my local dev machine?~ (nope I misread)

Can you please do a `bundle show faraday` and tell me if you are using faraday 1 or faraday 2 in your project?  And also if you're willing, a `bundle update` and see if tests still pass locally for you?

**Reply by jrochkind** [reply]
OK, I guessed wrong -- not fugit. That warning is actually coming from octokit. 

> /Users/jrochkind/.gem/ruby/3.0.4/gems/octokit-4.25.0/lib/octokit/default.rb:    Octokit::Warnable.octokit_warn 'To use retry middleware with Faraday v2.0+, install `faraday-retry` gem'

A dependency of `github_changelog_generator` (is that doing something useful for you?). 

It looks like it's something [octokit just outputs to error log](https://github.com/octokit/octokit.rb/blob/054a94ac4acf921ccd5a3dec77cb7f9508cc7001/lib/octokit/default.rb#L9-L15) on _boot_ (?!?), if you have faraday 2 loaded but don't have the faraday-retry middleware loaded. Which is annoying. And is messing up one of the tests when it gets output into stdout and  the test doesn't expect it there. 

But why doesn't this happen for you (still curious if you can make it happen for you with `bundle update`) or in CI? 

OK, well, CI is running via `appraisal` which I was not doing, I was just doing `./bin/rspec`.  (Can you please try running CI locally with just `./bin/rspec`? I bet you can reproduce some of this, especially after a `bundle update`!).   It's not obvious to me why that would make a difference, there's nothing in the Appraisal locking faraday to to 1.x or anything, but that might be it....

(btw why are you using a personal fork of `appraisal`? I have used appraisal before, fortunately!)

My just plain `./bin/rspec` looks like it was using Rails 6.1.6 for some reason (don't know why). So let's try `bundle exec appraisal install; bundle exec appraisal rails-6.1 rspec`

I still get the faraday-related error and three more (instead of 4?) locking-related errors. 

```
rspec ./spec/integration/server_spec.rb:6 # Server modes when development async successfully runs in the server
rspec ./spec/lib/good_job/lockable_spec.rb:124 # GoodJob::Lockable.with_advisory_lock opens a block with a lock that locks and unlocks records
rspec ./spec/lib/good_job/lockable_spec.rb:137 # GoodJob::Lockable.with_advisory_lock can unlock all advisory locks on the session with `unlock_session: true`
rspec ./spec/lib/good_job/lockable_spec.rb:100 # GoodJob::Lockable.advisory_lock_key when a block is passed locks that key for the bloc and then unlocks it
```

So it's not appraisal vs not, exactly. 

I'm going to leave for the day soon, not back until Monday. But I think it'd be really helpful if, when you had a chance, you confirmed whether or not any of these problems reproduce for you when you first `bundle update` and then `./bin/rspec` locally (not using appraisal).

**Reply by jrochkind** [reply]
Confirmed from [Github CI logs of a ruby 3.0 run](https://github.com/bensheldon/good_job/runs/7030032759?check_suite_focus=true) that it's using the same version of faraday (2.3.0) and octokit (4.25.0) as me -- why do I get an error that messes up the specs but CI doesn't?!? Odd.

**Reply by jrochkind** [reply]
Also, I misread, fugit/et-orbi are not involved with faraday at all!

**Reply by bensheldon** [reply]
Welp, I have the classic, `bin/setup && bin/rspec` works on my machine 😄 

<details>
 <summary>Shell output</summary>
 
```shell
❯ git clone git@github.com:bensheldon/good_job.git
Cloning into 'good_job'...
The authenticity of host 'github.com (192.30.255.112)' can't be established.
ED25519 key fingerprint is SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'github.com' (ED25519) to the list of known hosts.
remote: Enumerating objects: 6858, done.
remote: Counting objects: 100% (377/377), done.
remote: Compressing objects: 100% (298/298), done.
remote: Total 6858 (delta 100), reused 269 (delta 64), pack-reused 6481
Receiving objects: 100% (6858/6858), 4.68 MiB | 9.63 MiB/s, done.
Resolving deltas: 100% (3938/3938), done.
❯ cd good_job
❯ bin/setup

== Installing gem dependencies ==
Using rake 13.0.6
Using concurrent-ruby 1.1.10
Using i18n 1.10.0
Using minitest 5.15.0
Using tzinfo 2.0.4
Using zeitwerk 2.5.4
Using activesupport 6.1.5.1
Using builder 3.2.4
Using erubi 1.10.0
Using mini_portile2 2.6.1
Using racc 1.6.0
Using nokogiri 1.12.5
Using rails-dom-testing 2.0.3
Using crass 1.0.6
Using loofah 2.16.0
Using rails-html-sanitizer 1.4.2
Using actionview 6.1.5.1
Using rack 2.2.3
Using rack-test 1.1.0
Using actionpack 6.1.5.1
Using nio4r 2.5.8
Using websocket-extensions 0.1.5
Using websocket-driver 0.7.5
Using actioncable 6.1.5.1
Using globalid 1.0.0
Using activejob 6.1.5.1
Using activemodel 6.1.5.1
Using activerecord 6.1.5.1
Using marcel 1.0.2
Using mini_mime 1.1.2
Using activestorage 6.1.5.1
Using mail 2.7.1
Using actionmailbox 6.1.5.1
Using actionmailer 6.1.5.1
Using actiontext 6.1.5.1
Using pg 1.3.5
Using activerecord-explain-analyze 0.1.0
Using public_suffix 4.0.7
Using addressable 2.8.0
Using bundler 2.3.5
Using thor 1.2.1
Using appraisal 2.3.0 from https://github.com/bensheldon/appraisal.git (at fix-bundle-env@975180e)
Using ast 2.4.2
Using fiber-local 1.0.0
Using console 1.15.0
Using timers 4.3.3
Using async 1.30.1
Using async-io 1.33.0
Using async-pool 0.3.9
Using protocol-http 0.22.5
Using protocol-http1 0.14.2
Using protocol-hpack 1.4.2
Using protocol-http2 0.14.2
Using async-http 0.56.5
Using faraday-em_http 1.0.0
Using faraday-em_synchrony 1.0.0
Using faraday-excon 1.1.0
Using faraday-httpclient 1.0.1
Using multipart-post 2.1.1
Using faraday-multipart 1.0.3
Using faraday-net_http 1.0.1
Using faraday-net_http_persistent 1.2.0
Using faraday-patron 1.0.0
Using faraday-rack 1.0.0
Using faraday-retry 1.0.3
Using ruby2_keywords 0.0.5
Using faraday 1.10.0
Using async-http-faraday 0.11.0
Using benchmark-ips 2.10.0
Using html_tokenizer 0.0.7
Using parser 3.1.2.0
Using smart_properties 1.17.0
Using better_html 1.0.16
Using byebug 11.1.3
Using regexp_parser 2.3.1
Using xpath 3.2.0
Using capybara 3.35.3
Using chef-utils 17.10.0
Using childprocess 3.0.0
Using coderay 1.1.3
Using database_cleaner-core 2.0.1
Using database_cleaner-active_record 2.0.1
Using database_cleaner 2.0.1
Using diff-lcs 1.5.0
Using dotenv 2.7.6
Using thread 0.2.2
Using thread_safe 0.3.6
Using easy_translate 0.5.1
Using rainbow 3.1.1
Using parallel 1.22.1
Using rexml 3.2.5
Using rubocop-ast 1.17.0
Using ruby-progressbar 1.11.0
Using unicode-display_width 2.1.0
Using rubocop 1.28.2
Using erb_lint 0.1.1
Using et-orbi 1.2.7
Using faraday-http-cache 2.2.0
Using ffi 1.15.5
Using foreman 0.87.2
Using raabro 1.4.0
Using fugit 1.5.3
Using gem-release 2.2.2
Using multi_json 1.15.0
Using sawyer 0.8.2
Using octokit 4.22.0
Using github_changelog_generator 1.16.4
Using method_source 1.0.0
Using railties 6.1.5.1
Using webrick 1.7.0
Using good_job 2.17.0 from source at `.` and installing its executables
Using highline 2.0.3
Using rails-i18n 7.0.3
Using terminal-table 3.0.2
Using i18n-tasks 1.0.9
Using kramdown 2.4.0
Using kramdown-parser-gfm 1.1.0
Using matrix 0.4.2
Using mixlib-cli 2.1.8
Using tomlrb 2.0.1
Using mixlib-config 3.0.9
Using mixlib-shellout 3.2.7
Using mdl 0.11.0
Using memory_profiler 1.0.0
Using msgpack 1.5.1
Using optimist 3.0.1
Using pry 0.13.1
Using pry-byebug 3.9.0
Using pry-rails 0.3.9
Using puma 5.6.4
Using sprockets 4.0.3
Using sprockets-rails 3.4.2
Using rails 6.1.5.1
Using rbtrace 0.4.14
Using rspec-support 3.11.0
Using rspec-core 3.11.0
Using rspec-expectations 3.11.0
Using rspec-mocks 3.11.1
Using rspec-rails 5.1.2
Using rubocop-performance 1.13.3
Using rubocop-rails 2.14.2
Using rubocop-rspec 2.10.0
Using rubyzip 2.3.2
Using selenium-webdriver 3.142.7
Using sigdump 0.2.4
Using yard 0.9.27
Using yard-activesupport-concern 0.0.1
Bundle complete! 35 Gemfile dependencies, 147 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

== Setting up database ==
Created database 'good_job_development'
Created database 'good_job_test'
Inserted 1831 job records for a total of 1831 job records.

== Verifying system dependencies ==
ChromeDriver 102.0.5005.61 (0e59bcc00cc4985ce39ad31c150065f159d95ad3-refs/branch-heads/5005@{#819})
❯ bin/test_app_rake
Run options: --seed 17054

# Running:



Finished in 0.000286s, 0.0000 runs/s, 0.0000 assertions/s.
0 runs, 0 assertions, 0 failures, 0 errors, 0 skips
❯
❯ bin/rspec

Randomized with seed 12608
................................................................................................................DEPRECATION WARNING: GoodJob::Adapter's execution-related arguments (queues, max_threads, poll_interval, start_async_on_initialize) have been deprecated and will be removed in GoodJob v3. These options should be configured through GoodJob global configuration instead. (called from new at /Users/bensheldon/Repositories/tmp/good_job/spec/lib/good_job/adapter_spec.rb:95)
.....................................................................................................................DEPRECATION WARNING: GoodJob::Adapter's execution-related arguments (queues, max_threads, poll_interval, start_async_on_initialize) have been deprecated and will be removed in GoodJob v3. These options should be configured through GoodJob global configuration instead. (called from new at /Users/bensheldon/Repositories/tmp/good_job/spec/integration/adapter_spec.rb:70)
.DEPRECATION WARNING: GoodJob::Adapter's execution-related arguments (queues, max_threads, poll_interval, start_async_on_initialize) have been deprecated and will be removed in GoodJob v3. These options should be configured through GoodJob global configuration instead. (called from new at /Users/bensheldon/Repositories/tmp/good_job/spec/integration/adapter_spec.rb:70)
DEPRECATION WARNING: GoodJob::Adapter's execution-related arguments (queues, max_threads, poll_interval, start_async_on_initialize) have been deprecated and will be removed in GoodJob v3. These options should be configured through GoodJob global configuration instead. (called from new at /Users/bensheldon/Repositories/tmp/good_job/spec/integration/adapter_spec.rb:84)
..................................................................

Finished in 1 minute 48.2 seconds (files took 1.74 seconds to load)
296 examples, 0 failures

Randomized with seed 12608
```
</details>

Looking into the test failures you're having, I could see the potential for failures of test isolation in the environment. I did some improvements in https://github.com/bensheldon/good_job/pull/632

**Reply by jrochkind** [reply]
OK, weirdly, when I start over again and follow your script... I get tests passing now!  I don't know what the difference is, unless it's #652. 

I _do_ get `To use retry middleware with Faraday v2.0+, install `faraday-retry` gem` in my output though. Can't think of any reason that would be different for me and you.  It is just a warning not a problem, but an annoying one if I were getting it in my actual app logs....

Also, I notice you have `Gemfile.lock` checked into repo. I have taken the opposite practice, [for years](https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/).  

On the one hand, what you have done  does ensure that you and I are using exactly the same dependencies when running tests -- so isn't causing any problem here, and in fact may be avoiding problems.  But the downside is when we run `bin/rspec` we may be testing against old versions of some dependencies. You can see which ones by running `bundle outdated`.  For instance, bin/rspec still tests against Rails 6.1 and not Rails 7; and older versions of other dependencies than users may be using. However, you do have the appraisal individual gemfile locks in .gitignore, so I guess CI build is still testing against latest dependencies, so that's good...  I find the situation confusing, but perhaps you prefer it this way! 

So okay, for whatever mysterious reasons, tests are passing for me.... are you interested in me submitting PR's for changes to SQL ORDER in dequeing, for FIFO and/or queue priority?  And if so, any suggestions as to the best place and method to put a test for that behavior?

**Reply by bensheldon** [reply]
I am pretty sure it's #632. In the tests I was counting advisory locks across _all_ databases, not solely the test database 😓 

Yep, I'd gladly accept a PR if you wanted to submit for different dequeuing strategies. Probably should have a global configuration to allow the developer to choose a strategy.

I'm really appreciative of contributions, and please don't feel like you need to get everything into a single PR. We could do one additional strategy, get that reviewed and released, and then tackle the others. 🛹 ➡ 🚲 ➡ 🚗 ➡ ✈️ ➡ 🚀

**Reply by jrochkind** [reply]
OK, thanks, I'll see if I can find time to. 

I personally lean to think a totally generic "configure your own dequeing strategy" is overly complicated -- and I'm not sure how to do it; you'd configure a block that added a scope or something? I'm not totally sure that would even be sufficient for every dequeing strategy I know about (for instance, sidekiq's "weighted" queues) -- and I don't actually know about/can't think of very many!

To begin with, I my current thoughts personally are that FIFO/created_at should probably just be default/always -- unless there are known/demonstrated performance problems (which can' be solved with indexing), in which case it should just be binary configurable on/off.

**Reply by bensheldon** [reply]
I was imagining there would be several options preconfigured e.g.

```ruby
config.good_job.dequeuing_strategy = :fifo # or :random, or :queue_ordered
```

But, in re-reading #624, if you just want to change the existing default to be FIFO, I think we should just do that and defer the configuration until there are other other options.
